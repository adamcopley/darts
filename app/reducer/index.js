import { Alert } from 'react-native'
import { handleActions } from 'redux-actions'

import {
    SCORE_CHANGE,
    SCORE_CLEAR,
    SCORE_SUBMIT,
    SCORE_UNDO,
    GAME_START,
    GAME_START_NEW,
    GAME_SET_NAME,
    GAME_SET_TOTAL
} from '../constants'

export let initialState = {
    view: 'menu',
    error: '',
    total: 301,
    currentScore: '',
    currentPlayer: 'p1',
    p1: {score: 301, name: 'p1'},
    p2: {score: 301, name: 'p2'},
    history: []
}

export default handleActions({
    [SCORE_CHANGE]: (state, action) => {
        let prevScore = Number(state.currentScore)

        if (prevScore + action.payload <= 180) {
            return {
                ...state,
                currentScore: state.currentScore + String(action.payload)
            }
        }

        return {...state, error: 'The possible  highest score is 180'}
    },

    [SCORE_SUBMIT]: state => {
        let currentPlayer = state.currentPlayer
        let nextScore = state[currentPlayer].score - Number(state.currentScore)

        if (!state.currentScore) {
            return state
        }

        if (nextScore === 0 ) {
            Alert.alert(
                'Мы определили победителя!',
                `Победитель ${currentPlayer}`,
                [{text: 'OK'}],
                { cancelable: false }
            )

            return ({
                ...initialState,
                p1: {...initialState.p1, name: state.p1.name},
                p2: {...initialState.p2, name: state.p2.name}
            })
        }

        if (nextScore < 0) {
            return {
                ...state,
                history: [...state.history, 'Bust'],
                currentPlayer: currentPlayer === 'p1' ? 'p2' : 'p1'
            }
        }

        return {
            ...state,
            [currentPlayer]: {...state[currentPlayer], score: nextScore},
            history: [...state.history, state.currentScore],
            currentPlayer: currentPlayer === 'p1' ? 'p2' : 'p1'
        }
    },

    [SCORE_CLEAR]: state => ({
        ...state,
        error: '',
        currentScore: ''
    }),

    [SCORE_UNDO]: state => {
        if (state.currentScore) {
            return {...state, currentScore: '', error: ''}
        }

        let {history} = state

        if (history.length === 0) {
            return state
        }

        let lastDraw = history[history.length - 1]

        if (lastDraw === 'Bust') {
            lastDraw = 0
        } else {
            lastDraw = Number(lastDraw)
        }

        let prevPlayer = state.currentPlayer === 'p1' ? 'p2' : 'p1'

        return {
            ...state,
            history: history.slice(0, history.length - 1),
            [prevPlayer]: {...state[prevPlayer], score: state[prevPlayer].score + lastDraw},
            currentPlayer: prevPlayer
        }
    },

    [GAME_START]: state => ({
        ...state,
        view: 'game',
        p1: {...state.p1, score: state.total},
        p2: {...state.p2, score: state.total}
    }),

    [GAME_START_NEW]: state => ({
        ...initialState,
        p1: {...initialState.p1, name: state.p1.name},
        p2: {...initialState.p2, name: state.p2.name}
    }),

    [GAME_SET_NAME]: (state, {payload: {player, name}}) => ({
        ...state,
        [player]: {...state[player], name}
    }),

    [GAME_SET_TOTAL]: (state, {payload}) => ({
        ...state,
        total: payload
    })
}, initialState)
