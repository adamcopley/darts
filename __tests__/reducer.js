import {
    onScoreChange,
    onScoreSubmit,
    onScoreClear,
    onScoreUndo,
    onGameStart,
    onGameSetName,
    onGameSetTotal,
    onGameStartNew
} from '../app/actions'

import reducer, {initialState} from '../app/reducer'

describe('Actions snapshots', () => {
    test('onScoreChange under 180', () => {
        let nextState = reducer(initialState, onScoreChange(12))
        expect(nextState).toMatchSnapshot()
    })

    test('onScore change over 180', () => {
        let nextState = reducer(initialState, onScoreChange(181))
        expect(nextState).toMatchSnapshot()
    })

    test('onScoreSubmit empty score submit', () => {
        let nextState = reducer(initialState, onScoreSubmit())
        expect(nextState).toMatchSnapshot()
    })

    test('onScoreSubmit winning draw', () => {
        let nextState = reducer({
            ...initialState,
            p1: {score: 101},
            currentPlayer: 'p1',
            currentScore: 101
        }, onScoreSubmit())

        expect(nextState).toMatchSnapshot()
    })

    test('onScoreSubmit bust draw', () => {
        let nextState = reducer({
            ...initialState,
            p1: {score: 11},
            currentPlayer: 'p1',
            currentScore: 20
        }, onScoreSubmit())

        expect(nextState).toMatchSnapshot()
    })

    test('onScoreSubmit scored 1', () => {
        let nextState = reducer({
            ...initialState,
            p1: {score: 11},
            currentPlayer: 'p1',
            currentScore: 1
        }, onScoreSubmit())

        expect(nextState).toMatchSnapshot()
    })

    test('onScoreClear', () => {
        let nextState = reducer({
            ...initialState,
            error: 'Test error',
            currentScore: 101
        }, onScoreClear())

        expect(nextState).toMatchSnapshot()
    })

    test('onScoreUndo', () => {
        let nextState = reducer({
            ...initialState,
            error: 'Test error',
            currentScore: 101
        }, onScoreUndo())

        expect(nextState).toMatchSnapshot()
    })

    test('onGameStart', () => {
        let nextState = reducer(initialState, onGameStart())

        expect(nextState).toMatchSnapshot()
    })

    test('onGameStartNew', () => {
        let nextState = reducer({
            ...initialState,
            p1: {...initialState.p1, name: 'p1Name'},
            p2: {...initialState.p2, name: 'p2Name'}
        }, onGameStartNew())

        expect(onGameStartNew).toMatchSnapshot()
    })

    test('onGameSetName', () => {
        let nextState = reducer(initialState, onGameSetName('p1', 'test'))

        expect(nextState).toMatchSnapshot()
    })

    test('onGameSetTotal', () => {
        let nextState = reducer(initialState, onGameSetTotal('101'))

        expect(nextState).toMatchSnapshot()
    })
})