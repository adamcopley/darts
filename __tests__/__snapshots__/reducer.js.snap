// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions snapshots onGameSetName 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "",
  "error": "",
  "history": Array [],
  "p1": Object {
    "name": "test",
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;

exports[`Actions snapshots onGameSetTotal 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "",
  "error": "",
  "history": Array [],
  "p1": Object {
    "name": "p1",
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": "101",
  "view": "menu",
}
`;

exports[`Actions snapshots onGameStart 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "",
  "error": "",
  "history": Array [],
  "p1": Object {
    "name": "p1",
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "game",
}
`;

exports[`Actions snapshots onGameStartNew 1`] = `[Function]`;

exports[`Actions snapshots onScore change over 180 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "",
  "error": "The possible  highest score is 180",
  "history": Array [],
  "p1": Object {
    "name": "p1",
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;

exports[`Actions snapshots onScoreChange under 180 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "12",
  "error": "",
  "history": Array [],
  "p1": Object {
    "name": "p1",
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;

exports[`Actions snapshots onScoreClear 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "",
  "error": "",
  "history": Array [],
  "p1": Object {
    "name": "p1",
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;

exports[`Actions snapshots onScoreSubmit bust draw 1`] = `
Object {
  "currentPlayer": "p2",
  "currentScore": 20,
  "error": "",
  "history": Array [
    "Bust",
  ],
  "p1": Object {
    "score": 11,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;

exports[`Actions snapshots onScoreSubmit empty score submit 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "",
  "error": "",
  "history": Array [],
  "p1": Object {
    "name": "p1",
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;

exports[`Actions snapshots onScoreSubmit scored 1 1`] = `
Object {
  "currentPlayer": "p2",
  "currentScore": 1,
  "error": "",
  "history": Array [
    1,
  ],
  "p1": Object {
    "score": 10,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;

exports[`Actions snapshots onScoreSubmit winning draw 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "",
  "error": "",
  "history": Array [],
  "p1": Object {
    "name": undefined,
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;

exports[`Actions snapshots onScoreUndo 1`] = `
Object {
  "currentPlayer": "p1",
  "currentScore": "",
  "error": "",
  "history": Array [],
  "p1": Object {
    "name": "p1",
    "score": 301,
  },
  "p2": Object {
    "name": "p2",
    "score": 301,
  },
  "total": 301,
  "view": "menu",
}
`;
